### PR TITLE
Schemaless metrics + additional metrics for things we care about

### PR DESCRIPTION
**Pull request by fjy**
This is based on the discussion in the dev forums:

https://groups.google.com/forum/#!topic/druid-development/weccuTb2W54

This does not yet include doc updates.

This will be a backwards incompatible change for metrics.

The following changes will exist:
indexer/segment/bytes->segment/added/bytes
indexer/segmentMoved/byte->segment/moved/bytes
indexer/segmentNuked/bytes->segment/nuked/bytes
indexer/time/run/millis->task/run/time
exec/backlog->segment/scan/pending
coordinator/*->segment/*
server/*->segment/*
events/thrownAway->ingest/events/thrownAway
events/unparseable->ingest/events/unparseable
rows/output->ingest/rows/ouput
persists/num->ingest/persists/count
persists/time->ingest/persists/time
persists/backPressure->ingest/persists/backPressure
failed/persists->ingest/persists/failed
failed/handoff->ingest/handoff/failed

request/time->query/time
broker: query/time->query/node/time
historical/rt: query/time->query/segment/time
query/wait->query/wait/time

query/intervalChunk/time
query/segmentAndCache/time

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentMetadataUpdateAction.java [hunk 1]
```diff
@@ -63,12 +63,12 @@ public Void perform(
 
     // Emit metrics
     final ServiceMetricEvent.Builder metricBuilder = new ServiceMetricEvent.Builder()
-        .setUser2(task.getDataSource())
-        .setUser4(task.getType());
+        .setDimension("dataSource", task.getDataSource())
+        .setDimension("taskType", task.getType());
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentMetadataUpdateAction.java]: Can these be set to a constant somewhere to limit the capacity for typos?

File: processing/src/main/java/io/druid/query/QueryMetricUtil.java [hunk 2]
```diff
@@ -47,23 +64,24 @@ public String apply(Interval input)
...
+            "context",
             jsonMapper.writeValueAsString(
                 query.getContext() == null
                 ? ImmutableMap.of()
                 : query.getContext()
             )
         )
-        .setUser7(remoteAddr)
-        .setUser8(query.getId());
+        .setDimension("remoteAddr", remoteAddr)
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/QueryMetricUtil.java]: can we spell out `remoteAddress`?

File: processing/src/main/java/io/druid/query/aggregation/CountAggregatorFactory.java [hunk 3]
```diff
@@ -115,7 +115,7 @@ public String getName()
   @Override
   public String getTypeName()
   {
-    return "float";
+    return "long";
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/aggregation/CountAggregatorFactory.java]: can we put this in a separate PR or commit?

File: processing/src/main/java/io/druid/query/QueryMetricUtil.java [hunk 4]
```diff
@@ -23,18 +23,35 @@
...
 /**
  */
 public class QueryMetricUtil
 {
-  public static <T> ServiceMetricEvent.Builder makeQueryTimeMetric(Query<T> query)
+  public static int findNumComplexAggs(List<AggregatorFactory> aggs)
+  {
+    int retVal = 0;
+    for (AggregatorFactory agg : aggs) {
+      // This needs to change when
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/QueryMetricUtil.java]: when what?

File: processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java [hunk 5]
```diff
@@ -120,9 +120,8 @@ public boolean apply(T input)
   public ServiceMetricEvent.Builder makeMetricBuilder(DataSourceMetadataQuery query)
   {
     return new ServiceMetricEvent.Builder()
-        .setUser2(DataSourceUtil.getMetricName(query.getDataSource()))
-        .setUser4(query.getType())
-        .setUser6("false");
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java]: Where did boolean go?

File: processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQueryQueryToolChest.java [hunk 6]
```diff
@@ -122,9 +122,8 @@ public boolean apply(T input)
   public ServiceMetricEvent.Builder makeMetricBuilder(TimeBoundaryQuery query)
   {
     return new ServiceMetricEvent.Builder()
-        .setUser2(DataSourceUtil.getMetricName(query.getDataSource()))
-        .setUser4(query.getType())
-        .setUser6("false");
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQueryQueryToolChest.java]: Boolean vanished?

**Comment by drcrallen** [comment]
There are already concerns about the size of Metrics data. Do you have any numbers on how/if this changes data sizes?

File: server/src/main/java/io/druid/client/CachingClusteredClient.java [hunk 7]
```diff
@@ -325,7 +332,20 @@ private void addSequencesFromServer(ArrayList<Pair<Interval, Sequence<T>>> listO
...
+                  new Function<Query<T>, ServiceMetricEvent.Builder>()
+                  {
+                    @Override
+                    public ServiceMetricEvent.Builder apply(@Nullable final Query<T> input)
+                    {
+                      return toolChest.makeMetricBuilder(input);
+                    }
+                  },
+                  serverView.getQueryRunner(server),
+                  "segment/scan/time"
```
Comment (xvrl) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: this is not scanning a single segment

File: server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java [hunk 8]
```diff
@@ -261,7 +260,7 @@ public void onComplete(Result result)
       final long requestTime = System.currentTimeMillis() - start;
       try {
         emitter.emit(
-            QueryMetricUtil.makeRequestTimeMetric(jsonMapper, query, req.getRemoteAddr())
+            QueryMetricUtil.makeQueryTimeMetric(jsonMapper, query, req.getRemoteAddr())
                            .build("request/time", requestTime)
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/AsyncQueryForwardingServlet.java]: should this be "query/time" in line with the method name?

File: server/src/main/java/io/druid/server/coordination/ServerManager.java [hunk 9]
```diff
@@ -438,7 +438,7 @@ public void dropSegment(final DataSegment segment) throws SegmentLoadingExceptio
                           }
                         },
                         new ReferenceCountingSegmentQueryRunner<T>(factory, adapter),
-                        "scan/time"
+                        "segment/scan/time"
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/coordination/ServerManager.java]: maybe we should have all query related metrics start with `query`? e.g. `query/segment/time`, `query/cache/time` to be more consistent between all the query metrics

File: processing/src/main/java/io/druid/query/MetricsEmittingQueryRunner.java [hunk 10]
```diff
@@ -32,7 +32,7 @@
  */
 public class MetricsEmittingQueryRunner<T> implements QueryRunner<T>
 {
-  private static final String DEFAULT_METRIC_NAME = "query/time";
+  private static final String DEFAULT_METRIC_NAME = "query/portion/time";
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/MetricsEmittingQueryRunner.java]: `portion` is a little bit vague, maybe we can be a bit more specific? Maybe `query/historical/time` and `query/realtime/time`?

File: processing/src/main/java/io/druid/query/DruidMetrics.java [hunk 11]
```diff
@@ -23,18 +23,47 @@
...
+  public final static String TASK_STATUS = "taskStatus";
+
+  public final static String SERVER = "server";
+
+  public static int findNumComplexAggs(List<AggregatorFactory> aggs)
+  {
+    int retVal = 0;
+    for (AggregatorFactory agg : aggs) {
+      // This needs to change when we have support column types better
+      if (!agg.getTypeName().equals("float") && !agg.getTypeName().equals("long")) {
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/DruidMetrics.java]: HistogramAggregatorFactory throws UnsupportedOperationException on getTypeName(), 
need to check other aggregator factories too.

**Comment by xvrl** [comment]
@fjy would you mind also submitting a corresponding PR to https://github.com/metamx/server-metrics, to make SysMonitor and JvmMonitor schemaless as well?

File: indexing-service/src/main/java/io/druid/indexing/overlord/TaskQueue.java [hunk 12]
```diff
@@ -458,7 +465,7 @@ private void handleStatus(final TaskStatus status)
 
               // Emit event and log, if the task is done
               if (status.isComplete()) {
-                metricBuilder.setUser3(status.getStatusCode().toString());
+                metricBuilder.setDimension(DruidMetrics.TASK_STATUS, status.getStatusCode().toString());
                 emitter.emit(metricBuilder.build("indexer/time/run/millis", status.getDuration()));
```
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/overlord/TaskQueue.java]: task/runTime/millis

File: indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentInsertAction.java [hunk 13]
```diff
@@ -83,12 +84,12 @@ public SegmentInsertAction withAllowOlderVersions(boolean _allowOlderVersions)
...
-        .setUser2(task.getDataSource())
-        .setUser4(task.getType());
+        .setDimension(DruidMetrics.DATASOURCE, task.getDataSource())
+        .setDimension(DruidMetrics.TASK_TYPE, task.getType());
 
     for (DataSegment segment : segments) {
-      metricBuilder.setUser5(segment.getInterval().toString());
-      toolbox.getEmitter().emit(metricBuilder.build("indexer/segment/bytes", segment.getSize()));
+      metricBuilder.setDimension(DruidMetrics.INTERVAL, segment.getInterval().toString());
+      toolbox.getEmitter().emit(metricBuilder.build("segment/bytes", segment.getSize()));
```
Comment (fjy) [inline indexing-service/src/main/java/io/druid/indexing/common/actions/SegmentInsertAction.java]: segment/added/bytes

File: processing/src/main/java/io/druid/query/IntervalChunkingQueryRunner.java [hunk 14]
```diff
@@ -96,25 +99,30 @@ public IntervalChunkingQueryRunner(QueryRunner<T> baseRunner, QueryToolChest<T,
                     return toolChest.makeMetricBuilder(input);
                   }
                 },
-                baseRunner
-            ).withWaitMeasuredFromNow()),
-        executor, queryWatcher);
+                baseRunner,
+                "query/intervalChunk/time"
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/IntervalChunkingQueryRunner.java]: time/millis

File: server/src/main/java/io/druid/segment/realtime/RealtimeMetricsMonitor.java [hunk 15]
```diff
@@ -55,30 +56,30 @@ public boolean doMonitor(ServiceEmitter emitter)
...
       }
-      emitter.emit(builder.build("events/unparseable", unparseable));
-      emitter.emit(builder.build("events/processed", metrics.processed() - previous.processed()));
-      emitter.emit(builder.build("rows/output", metrics.rowOutput() - previous.rowOutput()));
-      emitter.emit(builder.build("persists/num", metrics.numPersists() - previous.numPersists()));
-      emitter.emit(builder.build("persists/time", metrics.persistTimeMillis() - previous.persistTimeMillis()));
+      emitter.emit(builder.build("ingest/events/unparseable", unparseable));
+      emitter.emit(builder.build("ingest/events/processed", metrics.processed() - previous.processed()));
+      emitter.emit(builder.build("ingest/rows/output", metrics.rowOutput() - previous.rowOutput()));
+      emitter.emit(builder.build("ingest/persists/num", metrics.numPersists() - previous.numPersists()));
```
Comment (fjy) [inline server/src/main/java/io/druid/segment/realtime/RealtimeMetricsMonitor.java]: /count

File: processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java (hunk 5 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java]: it was moved to a more common place in DruidMetrics

File: processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQueryQueryToolChest.java (hunk 6 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQueryQueryToolChest.java]: same comment as above

File: processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java (hunk 5 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/datasourcemetadata/DataSourceQueryQueryToolChest.java]: this information is not really needed in the new schema. These queries always have no filters

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 16]
```diff
@@ -193,7 +201,7 @@ public String apply(@Nullable AggregatorFactory input)
                         + prunedAggs.size()
                         + 1
                     );
-
+                    // JVM couldn't optimize this too well, so this is helping it out a bit.
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: I think @cheddar had removed this comment in a recent PR, maybe merge conflict?

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java [hunk 17]
```diff
@@ -253,7 +261,8 @@ public String apply(@Nullable AggregatorFactory input)
                         + query.getPostAggregatorSpecs().size()
                         + 1
                     );
-
+                    // put non finalized aggregators for calculating dependent post Aggregators
+                    // JVM is dumb about optimization
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: same about this comment

File: server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java [hunk 18]
```diff
@@ -53,29 +54,42 @@ public boolean doMonitor(ServiceEmitter emitter)
...
+                                        .setDimension(
+                                            "priority",
+                                            String.valueOf(serverConfig.getPriority())
+                                        );
+    emitter.emit(builder.build("segment/totalUsed", totalUsed));
     final double totalUsedPercent = serverConfig.getMaxSize() == 0 ? 0 : totalUsed / (double) serverConfig.getMaxSize();
-    emitter.emit(builder.build("server/segment/totalUsedPercent", totalUsedPercent));
-    emitter.emit(builder.build("server/segment/totalCount", totalCount));
+    emitter.emit(builder.build("segment/totalUsedPercent", totalUsedPercent));
+    emitter.emit(builder.build("segment/totalCount", totalCount));
```
Comment (xvrl) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: elsewhere we end in `/count`, any reason we don't here?

Comment (xvrl) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: those seem to be the only metrics that have camelCase parts?

File: docs/content/Metrics.md [hunk 19]
```diff
@@ -3,11 +3,182 @@ layout: doc_page
...
+
+## Coordination
+
+These metrics are for the Druid coordinator and are reset each time the coordinator runs the coordination logic.
+
+The following metrics are emitted on a per tier basis:
+
+|Metric|Description|Normal Value|
+|--------|-----------|-------|
+|`segment/*/added/count`|Number of segments added to the cluster.|Varies.|
```
Comment (himanshug) [inline docs/content/Metrics.md]: can we put "tier" as a dimension instead of encoding it in the metric name as is done for some other metrics?

File: processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java (hunk 16 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/topn/TopNQueryQueryToolChest.java]: yeah, this is a merge conflict, i'll remove it

File: server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java (hunk 18 above)
Comment (fjy) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: we also use it in some other places (intervalChunk)

Comment (fjy) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: maybe we should make a decision about that

File: processing/src/main/java/io/druid/query/DruidMetrics.java (hunk 11 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/query/DruidMetrics.java]: we should probably fix this before merging

File: server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java (hunk 18 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: ok, it's up to you, since you came up with the conventions, but I like things being consistent.

**Comment by xvrl** [comment]
I'm happy to merge this as long as:
- [ ] everyone is ok with making the backwards incompatible change in metrics. @cheddar, thoughts?
- [ ] remaining issues with exceptions being thrown are resolved

File: docs/content/Metrics.md [hunk 20]
```diff
@@ -3,11 +3,182 @@ layout: doc_page
...
+
+All Druid metrics share a common set of fields:
+
+* `timestamp` - the time the metric was created
+* `metric` - the name of the metric
+* `service` - the service name that emitted the metric
+* `host` - the host name that emitted the metric
+* `value` - some numeric value associated with the metric
+
+Each metric also has their own set of dimensions that vary from metric to metric.
```
Comment (himanshug) [inline docs/content/Metrics.md]: can we document the dimensions as well?

File: docs/content/Metrics.md [hunk 21]
```diff
@@ -3,11 +3,182 @@ layout: doc_page
...
+|`segment/loadQueue/size`|Size in bytes of segments to load.|Varies.|
+|`segment/loadQueue/failed`|Number of segments that failed to load.|0|
+|`segment/loadQueue/count`|Number of segments to load.|Varies.|
+|`segment/dropQueue/count`|Number of segments to drop.|Varies.|
+
+The following metrics are emitted on a per datasource basis:
+
+|Metric|Description|Normal Value|
+|--------|-----------|-------|
+|`segment/size`|Maximum byte limit available for segments.|Varies.|
```
Comment (himanshug) [inline docs/content/Metrics.md]: is this the "total byte size of segments" in the datasource?

File: docs/content/Metrics.md [hunk 22]
```diff
@@ -3,11 +3,182 @@ layout: doc_page
...
+|`segment/count`|Bytes used for served segments.|< max|
```
Comment (himanshug) [inline docs/content/Metrics.md]: this is supposed to be "number of served segments" instead of "bytes used..."

File: server/src/main/java/io/druid/client/CachingClusteredClient.java [hunk 23]
```diff
@@ -325,7 +332,20 @@ private void addSequencesFromServer(ArrayList<Pair<Interval, Sequence<T>>> listO
...
+                  new Function<Query<T>, ServiceMetricEvent.Builder>()
+                  {
+                    @Override
+                    public ServiceMetricEvent.Builder apply(@Nullable final Query<T> input)
+                    {
+                      return toolChest.makeMetricBuilder(input);
+                    }
+                  },
+                  serverView.getQueryRunner(server),
+                  "query/node/time"
```
Comment (himanshug) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: this is a good metric to have but it will be even more useful if metric event includes the name of historical/realtime node as a dimension. so, when I see high/abnormal value of this metric, I would immediately know which historical/realtime node is causing trouble.

File: docs/content/Metrics.md (hunk 20 above)
Comment (fjy) [inline docs/content/Metrics.md]: i was hoping to do this in later PRs and make progress here first. The list of dimensions can wildly vary from metric to metric.

File: processing/src/main/java/io/druid/query/DruidMetrics.java (hunk 11 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/DruidMetrics.java]: i fixed that already

File: docs/content/Metrics.md (hunk 19 above)
Comment (fjy) [inline docs/content/Metrics.md]: yeah, that's a good point

File: docs/content/Metrics.md (hunk 21 above)
Comment (fjy) [inline docs/content/Metrics.md]: no, it is the maxSize configured for the historical node

File: docs/content/Metrics.md (hunk 22 above)
Comment (fjy) [inline docs/content/Metrics.md]: fixed.

File: docs/content/Metrics.md (hunk 19 above)
Comment (fjy) [inline docs/content/Metrics.md]: added.

File: docs/content/Metrics.md (hunk 20 above)
Comment (fjy) [inline docs/content/Metrics.md]: actually i just ended up doign it now

File: server/src/main/java/io/druid/client/CachingClusteredClient.java (hunk 23 above)
Comment (fjy) [inline server/src/main/java/io/druid/client/CachingClusteredClient.java]: added server as a dimension

File: server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java (hunk 18 above)
Comment (fjy) [inline server/src/main/java/io/druid/server/metrics/HistoricalMetricsMonitor.java]: actually total isn't needed. Looked through the code and removed it

**Comment by fjy** [comment]
@xvrl @himanshug Addressed all code review comments.

**Comment by fjy** [comment]
@xvrl @himanshug Any thoughts?

**Comment by xvrl** [comment]
LGTM, @himanshug any other comments?

File: docs/content/Metrics.md [hunk 24]
```diff
@@ -3,11 +3,168 @@ layout: doc_page
...
+|`segment/deleted/count`|Number of segments dropped due to rules.|tier.|Varies.|
+|`segment/unneeded/count`|Number of segments dropped due to being marked as unused.|tier.|Varies.|
+|`segment/cost/raw`|Used in cost balancing. The raw cost of hosting segments.|tier.|Varies.|
+|`segment/cost/normalization`|Used in cost balancing. The normalization of hosting segments.|tier.|Varies.|
+|`segment/cost/normalized`|Used in cost balancing. The normalized cost of hosting segments.|tier.|Varies.|
+|`segment/loadQueue/size`|Size in bytes of segments to load.|server.|Varies.|
+|`segment/loadQueue/failed`|Number of segments that failed to load.|server.|0|
+|`segment/loadQueue/count`|Number of segments to load.|server.|Varies.|
+|`segment/dropQueue/count`|Number of segments to drop.|server.|Varies.|
+|`segment/size`|Maximum byte limit available for segments.|dataSource.|Varies.|
```
Comment (himanshug) [inline docs/content/Metrics.md]: himanshu: is this the "total byte size of segments" in the datasource?
fj: no, it is the maxSize configured for the historical node

himanshu: historical has maxSize configured globally (not per datasource level), how come does this metric has dataSource as a dimension then? also , code at https://github.com/druid-io/druid/pull/1208/files#diff-5494ef6d4623daa4fc35a5558f023294L216 seems to suggest its just total size of segments for the data source.

Comment (fjy) [inline docs/content/Metrics.md]: My bad, for some reason I thought you were commenting on 'segment/max'. Will fix

**Comment by himanshug** [comment]
LGTM

**Comment by xvrl** [comment]
Looks like we need to update docs / configs to rename `ServerMonitor` to `HistoricalMetricsMonitor`
